recipes/ppo_full_finetune_single_device.py

```diff
@@ -0,0 +1,1139 @@
+                shape [b, max_generated_tokens]
+            - value_padding_masks (torch.Tensor): padding masks for the values with
+                shape [b, max_generated_tokens]
+            - value_seq_idxs (torch.Tensor): indexes of the token
+                after the last valid (non-padding) token in the responses with shape [b]
+            - scores (torch.Tensor): scores from the reward model with shape [b]
+
+        """
+        batch_size, context_length = input_ids.shape
+
+        # step 1: generate responses, and logits corresponding to the responses using the policy
+        query_responses, logits = rlhf.generate_with_logits(
+            model=self._policy_model,
+            prompt=input_ids,
+            max_generated_tokens=self._max_generated_tokens,
+            temperature=self._temperature,
+            top_k=self._top_k,
+            pad_id=self._tokenizer.pad_id,
+            rng=self._rng,
+        )
+
+        responses = query_responses[:, context_length:].clone()
+        query_response_padding_masks = query_responses == self._tokenizer.pad_id
+
+        # step 1.1 create attention masks and position IDs for any padding tokens in inputs, used for future forward passes
+        masks = rlhf.get_causal_mask(~(query_response_padding_masks))
+        position_ids = (~query_response_padding_masks).cumsum(-1) - (
+            ~query_response_padding_masks
+        ).long()
+        position_ids = position_ids.type(torch.int)
+
+        del query_response_padding_masks
+
+        # step 2. estimate logprobs of the generated responses using the current policy
+        logits = logits[:, context_length - 1 :]
+        logprobs = rlhf.logits_to_logprobs(logits, responses, self._temperature)
+
+        del logits
+
+        # step 2.1 estimate logprobs of the generated responses using the reference policy
+        ref_logits = self._ref_policy_model(
+            query_responses, input_pos=position_ids, mask=masks
+        )
+        ref_logits = rlhf.query_response_logits_to_response_logits(
+            ref_logits, context_length
+        )
+        ref_logprobs = rlhf.logits_to_logprobs(ref_logits, responses, self._temperature)
+
+        del ref_logits
+
+        # step 3. estimate values from the generated responses using the value function
+        values = self._value_model(query_responses, input_pos=position_ids, mask=masks)
+        values = rlhf.query_response_logits_to_response_logits(
+            values, context_length
+        ).squeeze(-1)
+
+        # step 4. replace any tokens in the response after the first stop token (usually EOS token) with padding
+        response_padding_masks, responses = rlhf.truncate_sequence_at_first_stop_token(
+            responses, self._stop_token_ids, self._tokenizer.pad_id
+        )
+
+        # step 5. run the reward model on the query truncated-response pairs
```

Comment:
nit: all the places you say stuff like "query truncated-response pair" or "query-generated-response pair" I would say it like "(query, truncated response) pair" or "(query, generated response) pair" to make it explicit which words map to distinct entities